icicle/tests/test_field_api.cpp

```diff
@@ -939,6 +942,128 @@ TYPED_TEST(FieldApiTest, ntt)
+}
+
+TEST_F(FieldApiTestBase, CpuProgramExecutor)
+{
+  int seed = time(0);
+  srand(seed);
+  ICICLE_LOG_DEBUG << "seed = " << seed;
+
+  // randomize input vectors
+  const int total_size = 100000;
+  auto in_a = std::make_unique<scalar_t[]>(total_size);
+  scalar_t::rand_host_many(in_a.get(), total_size);
+  auto in_b = std::make_unique<scalar_t[]>(total_size);
+  scalar_t::rand_host_many(in_b.get(), total_size);
+  auto in_c = std::make_unique<scalar_t[]>(total_size);
+  scalar_t::rand_host_many(in_c.get(), total_size);
+  auto in_eq = std::make_unique<scalar_t[]>(total_size);
+  scalar_t::rand_host_many(in_eq.get(), total_size);
+
+  //----- element wise operation ----------------------
+  auto out_element_wise = std::make_unique<scalar_t[]>(total_size);
+  std::ostringstream oss;
+  oss << "Strait forward function (Element wise) time: ";
+
+  START_TIMER(element_wise_op)
+  for (int i = 0; i < 100000; ++i) {
+    out_element_wise[i] = in_eq[i] * (in_a[i] * in_b[i] - in_c[i]);
+  }
+  // std::cout << "Strait forward function (Element wise) time: ";
+  END_TIMER(element_wise_op, oss.str().c_str(), true);
+
+  //----- written program ----------------------
+  Program<scalar_t> program_written(&combine_func, 4);
+  // program_written.print_program();
+
+  CpuProgramExecutor<scalar_t> prog_exe_written(program_written);
+  auto out_written_program = std::make_unique<scalar_t[]>(total_size);
+
+  // init program
+  prog_exe_written.m_variable_ptrs[0] = in_a.get();
+  prog_exe_written.m_variable_ptrs[1] = in_b.get();
+  prog_exe_written.m_variable_ptrs[2] = in_c.get();
+  prog_exe_written.m_variable_ptrs[3] = in_eq.get();
+  prog_exe_written.m_variable_ptrs[4] = out_written_program.get();
+
+  oss.str("");
+  oss.clear();
+  oss << "Program executor time: ";
+  // run on all vectors
+  START_TIMER(written_program)
+  for (int i = 0; i < total_size; ++i) {
+    prog_exe_written.execute();
+    (prog_exe_written.m_variable_ptrs[0])++;
+    (prog_exe_written.m_variable_ptrs[1])++;
+    (prog_exe_written.m_variable_ptrs[2])++;
+    (prog_exe_written.m_variable_ptrs[3])++;
+    (prog_exe_written.m_variable_ptrs[4])++;
+  }
+  END_TIMER(written_program, oss.str().c_str(), true);
+
+  // check correctness
+  ASSERT_EQ(0, memcmp(out_element_wise.get(), out_written_program.get(), total_size * sizeof(scalar_t)));
+
+  //----- predefined program ----------------------
+  Program<scalar_t> predef_program(EQ_X_AB_MINUS_C);
+  // predef_program.print_program();
+
+  CpuProgramExecutor<scalar_t> prog_exe_predef(predef_program);
+  auto out_predef_program = std::make_unique<scalar_t[]>(total_size);
+
+  // init program
+  prog_exe_predef.m_variable_ptrs[0] = in_a.get();
+  prog_exe_predef.m_variable_ptrs[1] = in_b.get();
+  prog_exe_predef.m_variable_ptrs[2] = in_c.get();
+  prog_exe_predef.m_variable_ptrs[3] = in_eq.get();
+  prog_exe_predef.m_variable_ptrs[4] = out_predef_program.get();
+
+  oss.str("");
+  oss.clear();
+  oss << "Program predefined time: ";
```

Comment:
minor: note you can simply put this string directly in END_TIMER in this case

Reply:
its already there. 
I don't follow.

Reply:
I mean you don't need oss for a simple string but never mind